infinitory/errors.py

```diff
@@ -0,0 +1,137 @@
+
+        self._logger = logging.getLogger()
+
+    def delete_report_cache(self):
+        cache_path = './cache/'
+        for f in os.listdir(cache_path):
+            # Delete cache item if older than 2 hours
+            if os.stat(os.path.join(cache_path,f)).st_mtime < time.mktime(datetime.now().timetuple()) - (1 * 3600):
+                print("Deleting File " + os.path.join(cache_path,f))
+                os.remove(os.path.join(cache_path,f))
+
+    def load_reports_fast(self, pupdb):
+        # Takes Much Longer
+        self.reports = dict()
+        for report in pupdb.query('reports[] { hash in nodes[latest_report_hash] { } }'):
+            if report['status'] == 'failed':
+                self.reports[report['certname']] = report
+
+    def load_reports(self, pupdb):
+        for report in pupdb.query('nodes[certname, latest_report_hash] { }'):
+            cache_file = "./cache/%s" % report["latest_report_hash"]
+            if os.path.isfile(cache_file):
+                full_report = pickle.load(open(cache_file, "rb"))
+                sys.stdout.write('#')
+            else:
+                full_report = pupdb.query('reports[] { hash = "' + report["latest_report_hash"] + '" }')
+                pickle.dump( full_report, open(cache_file, "wb" ) )
+                sys.stdout.write('.')
+            sys.stdout.flush()
+
+            self._reports[report["certname"]] = full_report[0]
+
+    def stored_errors(self):
+        return [
+            "Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Function Call, Untrusted facts (left) don't match values from certname (right)"
+        ]
+
+    def matches_stored_error(self, message):
+        for se in self.stored_errors():
+            if message.startswith(se):
+                return True, se
+
+        return False, ""
+
+    def set_all_errors(self, input):
+        # Mainly used during testing to mock values
+        self._all_errors = input
+
+    def set_unique_errors(self, input):
+        # Mainly used during testing to mock values
+        self._unique_errors = input
+
+    def all_errors(self):
+        retval = []
+        for i, ue in enumerate(self._all_errors):
+            retval.append({
+                "other": ue
+            })
+        return retval
+
+    def unique_errors(self):
+        retval = []
+        for i, ue in enumerate(self._unique_errors):
+            retval.append({
+                "other": ue
+            })
+        return retval
+
+    def clean_error_message(self, error_message):
+        matched, stored_error = self.matches_stored_error(error_message)
+
+        if matched:
+            return stored_error
+
+        return error_message
+
+    def modify_unique_errors_at(self, i, log_level, hostname):
+
+        new_hostname_list = self._unique_errors[i]['hostnames']
+        new_hostname_list.append(hostname)
+
+        self._unique_errors[i] = {
```

Comment:
It's more idiomatic to modify the entry than to recreate it. [`defaultdict`](
https://docs.python.org/2/library/collections.html#collections.defaultdict) makes that particularly easy.

Reply:
Tell me more, I think I tried to do that but went this way, because the cellformatter expects a list.